src/query-executor/query-executor-base.ts

```diff
@@ -68,6 +69,35 @@ export abstract class QueryExecutorBase implements QueryExecutor {
     })
   }
 
+  async *stream<R>(
+    compiledQuery: CompiledQuery,
+    chunkSize: number,
+    queryId: QueryId
+  ): AsyncIterableIterator<QueryResult<R>> {
+    const connectionDefer = new Deferred<DatabaseConnection>()
+    const connectionReleaseDefer = new Deferred<void>()
+
+    this.provideConnection(async (connection) => {
+      connectionDefer.resolve(connection)
+
+      // Lets wait until we don't need connection before returning here (returning releases connection)
+      return await connectionReleaseDefer.promise
+    }).catch((ex) => connectionDefer.reject(ex))
+
+    const connection = await connectionDefer.promise
+
+    try {
+      for await (const result of connection.executeQueryStream(
+        compiledQuery,
+        chunkSize
+      )) {
+        yield await this.#transformResult(result, queryId)
```

Comment:
If the iteration is ended prematurely in the user code, this loop will get stuck forever and it will hold the connection forever. Right?

I think you need to implement the `AsyncIterator` interface "manually". Here's the interface

```ts
interface AsyncIterator<T, TReturn = any, TNext = undefined> {
    // NOTE: 'next' is defined using a tuple to ensure we report the correct assignability errors in all places.
    next(...args: [] | [TNext]): Promise<IteratorResult<T, TReturn>>;
    return?(value?: TReturn | PromiseLike<TReturn>): Promise<IteratorResult<T, TReturn>>;
    throw?(e?: any): Promise<IteratorResult<T, TReturn>>;
}
```

There you can implement the `return` and `throw` methods to release the connection. They get called when the iteration is ended by the consumer.

Since this is a really critical thing (we don't want to leak connections) we need a test for this. One way to test this is to run a code that breaks the iteration midway multiple times. To be specific, the code needs to be run `connectionPoolMaxSize + 1` times. You could create a constant for that in the tests. The same constant would be used to create the kysely instances AND in this test.

Reply:
Also I think you need to reserve the connection only when the `next` method is called the first time to make sure has been started. Otherwise the case where user creates a stream but never iterates it will leak a connection.

Reply:
Or if this method returns an `AsyncIterable`, you can reserve the connection when the `[Symbol.asyncIterator]()` method is called. I don't think there's a need to use `AsyncIterableIterator`? Why is it used here? Doesn't this method actually return a `AsyncGenerator`?

Reply:
This doesn't seem to be the case. Added a test case for it though. Good idea!

Reply:
You're right! I had to write some test code to understand how generators created using `async function *func()` behave. It seems that the code after the last `yield` is never executed BUT the code in any `finally` blocks is executed. And the `return` calls "bubble in" when the outermost loop ends. I learn't something new today!